### PR TITLE
转换规则 添加部分 inplace api 的映射 PART.01

### PR DESCRIPTION
**Pull request by RedContritio**
### PR Docs

- https://github.com/PaddlePaddle/docs/pull/6169
- https://github.com/PaddlePaddle/docs/pull/6189

### PR APIs

```bash
torch.nn.functional.hardtanh_
torch.nn.functional.leaky_relu_
torch.abs_
torch.Tensor.abs_ # 尚未支持，留置
torch.Tensor.bitwise_not_
torch.Tensor.bitwise_and_
torch.Tensor.bitwise_or_
torch.Tensor.bitwise_xor_
torch.Tensor.cumprod_
torch.Tensor.cumsum_
torch.Tensor.detach_             # 已移除，留在此后 PR 更新单测
torch.Tensor.digamma_
torch.Tensor.divide_              # 已移除，留在此后 PR 更新
torch.Tensor.eq_                   # 已移除，留在此后 PR 更新
torch.Tensor.floor_divide_
torch.Tensor.frac_
torch.Tensor.gcd_
torch.Tensor.ge_                   # 已移除，留在此后 PR 更新
torch.Tensor.greater_equal_ # 已移除，留在此后 PR 更新
torch.Tensor.gt_                   # 已移除，留在此后 PR 更新
torch.Tensor.greater_           # 已移除，留在此后 PR 更新
torch.Tensor.index_add_
torch.Tensor.index_put_
```

尚未解决的问题：

- `torch.Tensor.detach_` 的单测运行时会发生段错误，但转换前后的代码独立运行结果正常，目前判断是 `paddle` 主框架的问题，已 issue https://github.com/PaddlePaddle/Paddle/issues/57303
- `torch.Tensor.divide_` 的单测，已单独 pr #285。当注释掉 `api_matcher.py` 中以下两行时，会导致解析语法错误，疑似项目处理代码的逻辑存在 bug：生成 `a.divide_(...).floor_()` 后，会将 `.floor_()` 自动放到下一行：
```python
        # if remove this 2 lines, a grammer error will be raised
        if self.is_inplace_method():
            kwargs["out"] = self.paddleClass
```

### CODE REVIEW

**Review by zhwesky2010 — COMMENTED** [review]
新增单测需要按更高的规范要求来，请参考 https://github.com/PaddlePaddle/PaConvert/blob/master/docs/CONTRIBUTING.md#%E6%AD%A5%E9%AA%A45%E7%BC%96%E5%86%99%E5%8D%95%E5%85%83%E6%B5%8B%E8%AF%95

File: tests/test_Tensor_ge_.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+obj = APIBase("torch.ge_")
+
+
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        x = torch.tensor([[1, 2], [3, 4]])
+        x.ge_(torch.tensor([[1, 1], [4, 4]]))
```
Comment (zhwesky2010) [inline tests/test_Tensor_ge_.py:27]: 这种compare op，咱们和torch的输出类型应该不同吧，不知道怎么测过了，补充一下float的测试

File: paconvert/api_mapping.json [hunk 2]
```diff
@@ -3958,6 +4078,14 @@
       "input": "x"
     }
   },
+  "torch.cumprod_": {
+    "Matcher": "CumprodMatcher",
```
[the remark this answers is not in the document]
Comment (RedContritio) [reply paconvert/api_mapping.json]: 这个是写错了，已删除

File: tests/test_Tensor_ge_.py (hunk 1 above)
Comment (RedContritio) [reply tests/test_Tensor_ge_.py:27]: 测试了一下 float 的，输出结果的类型和值是一样的，我准备分别补两个单测作为验证

File: paconvert/api_mapping.json [hunk 3]
```diff
@@ -1128,7 +1227,18 @@
   "torch.Tensor.index_fill": {},
   "torch.Tensor.index_fill_": {},
   "torch.Tensor.index_put": {},
-  "torch.Tensor.index_put_": {},
+  "torch.Tensor.index_put_": {
+    "Matcher": "GenericMatcher",
+    "paddle_api": "paddle.Tensor.index_put_",
+    "args_list": [
+      "indices",
+      "value",
```
[the remark this answers is not in the document]
Comment (RedContritio) [reply paconvert/api_mapping.json:1244]: 这个看起来是 `args_list`  写的不对，但 `kwargs_change` 用的是正确版本，所以没有触发对应的检查。

[the remark this answers is not in the document]
Comment (RedContritio) [reply paconvert/api_mapping.json:1244]: 貌似是最开始就没有加入对应检查机制，此后还被移除掉了 "TODO" 标志。

https://github.com/PaddlePaddle/PaConvert/commit/f85e98847bb45b012e0739221cf5f526f81008a4#diff-7f0add6089340148b0b9e4ee1e5b7c495dd4554c1e45fa390161d3de83a84202R195-R196

[the remark this answers is not in the document]
Comment (zhwesky2010) [reply paconvert/api_mapping.json]: 目前完善了检查机制，后面写args_list需要完全覆盖测试

File: tests/test_Tensor_greater_equal_.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+# limitations under the License.
+
+import textwrap
+
+from apibase import APIBase
+
+obj = APIBase("torch.Tensor.greater_equal_")
+
+
+def test_case_1():
```
Comment (zhwesky2010) [inline tests/test_Tensor_greater_equal_.py:22]: compare op都有这个问题

**Review by zhwesky2010 — COMMENTED** [review]
还需要处理下冲突，另外就是目前的 `args_list` 写法上稍微有些变化，需要将 **强制指定关键字** 参数也标进来，例如：
```
    "args_list": [
      "input",
      "other",
      "*",
      "out"
    ]
```

同时对于 **所有类成员API**，需要增加一个：`min_input_args` 来表示最少输入参数个数，例如3个参数其中有2个是默认参数，则最少输入1个：`"min_input_args": 1,` ，很多inplace是Tensor类方法，涉及到这个重构，需要配置一下新字段

File: tests/test_Tensor_cumprod_.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def test_case_6():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        x = torch.tensor([[1.0, 1.0, 1.0],
+                        [2.0, 2.0, 2.0],
+                        [3.0, 3.0, 3.0]])
+        x.cumprod_(1, dtype=torch.float32)
```
[the remark this answers is not in the document]
Comment (RedContritio) [reply tests/test_Tensor_cumprod_.py:94]: dtype 不输入关键字的话会无法转换，因为这个类型签名是 `Tensor.cumprod_(dim: int, *, dtype)`

File: tests/test_Tensor_eq_.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+obj = APIBase("torch.Tensor.eq_")
+
+
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        x = torch.tensor([[1, 2], [3, 4]])
+        x.eq_(torch.tensor([[1, 1], [4, 4]]))
+        x = x.bool()
```
[the remark this answers is not in the document]
Comment (RedContritio) [reply tests/test_Tensor_eq_.py:28]: 如果这样的话，`x.eq_(y).cast_(x.dtype)` 链式执行的时候，`x.eq_(y)` 之后已经改变了类型，最后的 `_cast` 是无效的，我认为这个无法在单行代码内完成

File: tests/test_Tensor_cumprod_.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhwesky2010) [reply tests/test_Tensor_cumprod_.py:94]: infoflow 2023-09-19 11-55-29

这些文档的bug真多啊，测试了才知道

File: tests/test_Tensor_eq_.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zhwesky2010) [reply tests/test_Tensor_eq_.py]: cast_改回来类型就可以了吧

File: tests/test_Tensor_cumprod_.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
```
Comment (zhwesky2010) [inline tests/test_Tensor_cumprod_.py:94]: infoflow 2023-09-19 11-56-32

这些文档的bug真多啊，测试了才能发现问题

File: tests/test_Tensor_eq_.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (RedContritio) [reply tests/test_Tensor_eq_.py]: 这个流程是先运算再求值的，等运算到 `cast` 的时候，`x.dtype` 已经是 `bool` 了

```python
>>> x.less_equal_(paddle.to_tensor([[1,3],[4,4]])).cast_(x.dtype)
Tensor(shape=[2, 2], dtype=bool, place=Place(cpu), stop_gradient=True,
       [[True, True],
        [True, True]])
```

[the remark this answers is not in the document]
Comment (RedContritio) [reply tests/test_Tensor_eq_.py]: 已单独提出作为 [Part.cmp PR](https://github.com/PaddlePaddle/PaConvert/pull/294)，发现 `cast_` 仍然存在非预期行为

**Review by zhwesky2010 — DISMISSED** [review]
LGTM

**Comment by zhwesky2010** [comment]
@RedContritio 这个`torch.abs_` 还没有支持，所以没有转报错了

File: tests/test_Tensor_eq_.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zhwesky2010) [reply tests/test_Tensor_eq_.py]: 了解问题了

**Comment by RedContritio** [comment]
> @RedContritio 这个`torch.abs_` 还没有支持，所以没有转报错了

已更新，支持了该方法的转换
